File: scripts/run_US_states.jl
using StanSample, Distributions, Dates
using DataFrames, CSV

ENV["JULIA_CMDSTAN_HOME"] = expanduser("~/cmdstan") # Path for cmdstan

include("../src/rt-from-frequency-dynamics.jl") # Loading project module
using .rt_from_frequency_dynamics

# 1. Get data
seq_df = DataFrame(CSV.File("../data/location-variant-sequence-counts.tsv"))
cases_df = DataFrame(CSV.File("../data/location-case-counts.tsv"))

# 2. Get lifetimes
g, onset = get_standard_delays()

# 3. Create the model for Rt
n_days = length(unique(cases_df.date))
rt_model = SplineTrend(20, n_days, 4)
priors = [["LAS"]]
seed_L = 7
forecast_L = 0

# 4. Choose lineage model type
LM =  FixedLineageModel(g, onset, rt_model, priors, seed_L, forecast_L)

# 5. Make stan models using shared model type
function make_state_models(seq_df::DataFrame, cases_df::DataFrame, LM::LineageModel, model_dir::String)
  loc_names = unique(seq_df.location)

  # Get state specific data structures
  LDs = LineageData[]
  for state in loc_names
    push!(LDs, 
      LineageData(filter(x -> x.location == state, seq_df), filter(x -> x.location == state, cases_df))
    )
  end

  # Making model objects to be run
  MSs = ModelStan[]
  for (state, LD) in zip(loc_names, LDs)
    # Changing name to be read by stan
    _state = replace(state, " " => "_")
    push!(MSs, make_stan(LM, LD, "rt-lineages-$(_state)", "$model_dir/$(_state)"))
  end
  return LDs, MSs
end

model_name = "all-states-preprint"
model_dir = "../data/sims/$(model_name)"
LDs, MSs = make_state_models(seq_df, cases_df, LM, model_dir)

# 6. Run stan models
for MS in MSs
  run!(MS, num_warmup = 1000, num_samples=1000, num_chains=4, get_samples=false) # Make sure to save samples
end

# 7. Last thing would be to export the dataframes
# today = "11_28_21"
# function export_dataframes(MSs, LDs)
#   df_list = DataFrame[]
#   for (MS, LD) in zip(MSs, LDs)
#     push!(df_list, get_Rt_dataframe(MD, LD))
#   end
#   return reduce(hcat, df_list)
# end

# rt_df = export_dataframes(MSs, LDs)
# CSV.write("../data/sims/results/inferred_lineage_rts_$(model_name)_$(today).tsv", delim ="\t", rt_df)

# # 8. Also writing the growth advantages
# function export_growth_advantages(MSs, LDs)
#   df_list = DataFrame[]
#   for (MS, LD) in zip(MSs, LDs)
#     push!(df_list, get_growth_advantages_dataframe(MD, LD))
#   end
#   return reduce(hcat, df_list)
# end

# ga_df = export_growth_advantages(MSs, LDs)
# CSV.write("../data/sims/results/inferred_lineage_growth_advantage_$(model_name)_$(today).tsv", delim ="\t", rt_df)

File: scripts/run_US_states.py
import numpyro
from numpyro.infer.autoguide import AutoMultivariateNormal
import pandas as pd
from src.rt_from_frequency_dynamics import *

ps = DefaultAes.ps

def run_SVI(LD, optimizer, num_samples=1000, iters=100_000, name="test", path=".", save=True, export=True, load=False):       
    X = make_breakpoint_splines(LD.cases, 20)
    data = LD.make_numpyro_input(X)

    # Defining model
    g, delays = get_standard_delays()
    LM = FixedGrowthModel(g, delays, 7, 0)

    # Run SVI
    SVIH = SVIHandler(optimizer=optimizer)
    guide = AutoMultivariateNormal(LM.model)
    
    if load:
        SVIH.load_state(f"{path}/models/{name}_svi.p")
    
    loss = SVIH.fit(LM.model, guide, data, iters, log_each=0)
    print(f"Model {name} finished. Final loss: {loss}")
    
    if jnp.isnan(loss):
        return False
    if save:
        SVIH.save_state(f"{path}/models/{name}_svi.p")
    
    # Get samples
    samples = SVIH.predict(LM.model, guide, data, num_samples=num_samples)
    dataset = to_arviz(samples)
    
    if export:
        # Get dataframes
        R_dataframe = pd.DataFrame(get_R(dataset, LD, ps[:-1], name))
        ga_dataframe = pd.DataFrame(get_growth_advantage(dataset, LD, ps[:-1], name))
    
        R_dataframe.to_csv(f"{path}/Rt/Rt_{name}.csv", encoding='utf-8', index=False)
        ga_dataframe.to_csv(f"{path}/ga/ga_{name}.csv", encoding='utf-8', index=False)
    return True

def get_state_LD(rc, rs, loc):
    rc_l = rc[rc.location == loc].copy()
    rs_l = rs[rs.location==loc].copy()
    return LineageData(rc_l, rs_l)

def run_locations(rc, rs, locations, optimizer, **kwargs):
    n_locations = len(locations)
    sucesses = []
    for i, loc in enumerate(locations):
        LD = get_state_LD(raw_cases, raw_seq, loc)
        model_name = loc.replace(" ", "_")
        sucess = run_SVI(LD, optimizer, name=model_name, **kwargs)
        sucesses.append(sucess)
        print(f'Location {loc} finished ({i+1}/{n_locations}).')
    return locations[successes]

def combine_exports(rc, rs, locations, path):   
    rt_list = []
    ga_list = []
    for i, loc in enumerate(locations):
        _loc = loc.replace(" ", "_")
        rt_list.append(pd.read_csv(f"{path}/Rt/Rt_{_loc}.csv"))
        ga_list.append(pd.read_csv(f"{path}/ga/ga_{_loc}.csv"))
    return pd.concat(rt_list), pd.concat(ga_list)


if __name__ == "__main__":
    raw_cases = pd.read_csv("../../rt-from-frequency-dynamics/data/location-case-counts.tsv", sep="\t")
    raw_seq = pd.read_csv("../../rt-from-frequency-dynamics/data/location-variant-sequence-counts.tsv", sep="\t")

    locations = pd.unique(raw_seq.location)

    optimizer = numpyro.optim.Adam(step_size=1.5e-4)
    num_samples = 1
    iters = 200_000
    path = "./sims/all-states-preprint"
    save = True
    export = True
    load = False
    succeded = run_locations(raw_cases,
                             raw_seq,
                             locations,
                             optimizer,
                             num_samples=num_samples,
                             iters=iters,
                             path=path,
                             save=save,
                             export=export,
                             load=load)
